### PR TITLE
Add Cache-Control to avatar redirects

### PR DESCRIPTION
**Pull request by silverwind**
This does seem to do the trick to make the Avatar redirects cachable in Chrome.

In Firefox, it does not seem to work, thought and I found no way to suppress the requests to the original URLs, I even tried setting 301 or an Etag to no avail.

Related discussion in https://github.com/go-gitea/gitea/issues/16964.

### CODE REVIEW

File: routers/web/user/avatar.go [hunk 1]
```diff
@@ -13,10 +13,16 @@ import (
 
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/httpcache"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
 )
 
+func doRedirect(ctx *context.Context, location string) {
```
Comment (zeripath) [inline routers/web/user/avatar.go]: `cacheableRedirect` might be a better name

**Review by zeripath — APPROVED** [review]
small nit but otherwise lgtm

**Review by axifive — APPROVED** [review]

File: routers/web/user/avatar.go (hunk 1 above)
Comment (zeripath) [reply routers/web/user/avatar.go]:
```suggestion
func cacheableRedirect(ctx *context.Context, location string) {
```

File: routers/web/user/avatar.go [hunk 2]
```diff
@@ -43,7 +49,7 @@ func Avatar(ctx *context.Context) {
 		user = models.NewGhostUser()
 	}
 
-	ctx.Redirect(user.RealSizedAvatarLink(size))
+	doRedirect(ctx, user.RealSizedAvatarLink(size))
```
Comment (zeripath) [inline routers/web/user/avatar.go]:
```suggestion
	cacheableRedirect(ctx, user.RealSizedAvatarLink(size))
```

File: routers/web/user/avatar.go [hunk 3]
```diff
@@ -63,7 +69,7 @@ func AvatarByEmailHash(ctx *context.Context) {
 		return
 	}
 	if len(email) == 0 {
-		ctx.Redirect(models.DefaultAvatarLink())
+		doRedirect(ctx, models.DefaultAvatarLink())
```
Comment (zeripath) [inline routers/web/user/avatar.go]:
```suggestion
		cacheableRedirect(ctx, models.DefaultAvatarLink())
```

File: routers/web/user/avatar.go [hunk 4]
```diff
@@ -94,5 +100,5 @@ func AvatarByEmailHash(ctx *context.Context) {
 		}
 	}
 
-	ctx.Redirect(models.MakeFinalAvatarURL(avatarURL, size))
+	doRedirect(ctx, models.MakeFinalAvatarURL(avatarURL, size))
```
Comment (zeripath) [inline routers/web/user/avatar.go]:
```suggestion
	cacheableRedirect(ctx, models.MakeFinalAvatarURL(avatarURL, size))
```

**Review by 6543 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
The default `StaticCacheTime` is pretty long (6 hours)

A user won't see his new avatar in 6 hours if he did a change ......

Is it good for end users?

**Comment by silverwind** [comment]
We could shorten the default time a bit. Gravatar and GitHub use 5 minutes, for example.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> In Firefox, it does not seem to work...

it appears to work in *my* firefox (just a standard firefox developer edition) 🎉 
have a look:
image

**Comment by silverwind** [comment]
It appears you're not using external avatars, so you're not hitting the redirect case. It should show up once you enable gravatar for example.

**Comment by silverwind** [comment]
Nice. In my testing I could not get Firefox to cache the redirect for whatever reason, even tried a completely fresh profile.

**Pull request merged**